### PR TITLE
fix: add query semantics & CircuitBreaker

### PR DESCRIPTION
**Pull request by lvermeulen**

### CODE REVIEW

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex.csproj [hunk 1]
```diff
@@ -1,5 +1,8 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
   <Import Project="..\..\packages\Be.Vlaanderen.Basisregisters.Build.Pipeline\Content\Be.Vlaanderen.Basisregisters.Build.Pipeline.Settings.Library.props" />
-
+  <ItemGroup>
+    <PackageReference Include="AWSSDK.Extensions.NETCore.Setup" Version="3.7.2" />
```
Comment (emalfroy) [inline src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex.csproj:5]: Shouldn't this be added with paket?

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            _mutex = new DynamoDBMutex(client, settings);
+
+
+            _duration = duration;
+            ResourceName = resourceName;
+        }
+
+        public async Task Open()
+        {
+            _lockToken = await _mutex.AcquireLockAsync(ResourceName, _duration);
```
Comment (emalfroy) [inline src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs:29]: Do we need exception handling here if another (lambda) instance concurrently took a lock?

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex.csproj (hunk 1 above)
Comment (lvermeulen) [reply src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex.csproj:5]: shouldn't be there at all, will fix

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs (hunk 2 above)
Comment (lvermeulen) [reply src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs:29]: i don't think so, that's why there's a mutex underneath

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDBMutex.cs [hunk 3]
```diff
@@ -1,3 +1,5 @@
+using System.Threading;
```
Comment (ArneD) [inline src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDBMutex.cs:1]: Move inside namespace

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDBMutex.cs [hunk 4]
```diff
@@ -220,5 +222,25 @@ public async Task<bool> DoesTableExists()
...
+            };
+
+            var response = await table.GetItemAsync(doc, new GetItemOperationConfig
+            {
+                AttributesToGet = new List<string> { ColumnNames.LeaseExpiry },
+                ConsistentRead = true
+            }, cancellationToken);
+
+            return response is not null
+                && DateTimeOffset.TryParse(response[ColumnNames.LeaseExpiry], out var dateTimeOffset)
```
Comment (ArneD) [inline src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDBMutex.cs:242]: maybe rename `dateTimeOffset` to `lockExpirationDateTime` or something like that.

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        private LockToken? _lockToken;
+
+        public string ResourceName { get; }
+
+        public DynamoDbCircuitBreaker(string resourceName, TimeSpan duration, AmazonDynamoDBClient? dynamoDbClient = null)
+        {
+            var client = dynamoDbClient ?? new AmazonDynamoDBClient();
+            var settings = new DynamoDBMutexSettings { CreateTableIfNotExists = true };
+            _mutex = new DynamoDBMutex(client, settings);
+
```
Comment (ArneD) [inline src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs:21]: empty line

**Review by ArneD — CHANGES_REQUESTED** [review]

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDBMutex.cs (hunk 4 above)
Comment (lvermeulen) [reply src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDBMutex.cs:242]: fixed

File: src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs (hunk 5 above)
Comment (lvermeulen) [reply src/Be.Vlaanderen.Basisregisters.Aws.DistributedMutex/DynamoDbCircuitBreaker.cs:21]: fixed

**Pull request merged**